### PR TITLE
EOS-26811: Get object with range read is not working on multipart-uploaded object

### PR DESCRIPTION
**Pull request by DPG17**
Signed-off-by: Dattaprasad Govekar <dattaprasad.govekar@seagate.com>

# Problem Statement
- Range based Get object on multipart uploaded object fails when first byte offset of range does
not coincide with the start of the data block unit. The issue is seen only when max units read/write per request is more than one (say, 8 or 32)

# Design
-  Handled situation when the first byte range is not aligned to start of data block

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: server/s3_get_object_action.cc [hunk 1]
```diff
@@ -760,6 +760,26 @@ void S3GetObjectAction::read_object_data() {
     } else {
       blocks_to_read = total_blocks_to_read - blocks_already_read;
     }
+    // Below is only for first block read op and non-zero block start offset.
+    // For the first block read op, if reading starts from non-zero offset
+    // adjust number of 'blocks_to_read', so that the read does not go
+    // beyond the object size.
+    if (blocks_already_read == 0 && (motr_reader->get_last_index() != 0)) {
```
Comment (knrajnambiar76) [inline server/s3_get_object_action.cc:767]: please test with different scenarios
range falling within a part/across parts/within last part/range as complete part

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_get_object_action.cc (hunk 1 above)
Comment (DPG17) [reply server/s3_get_object_action.cc:784]: Tested scenarios (range falling within part, across parts, within last part) with 1 and 8 motr_units_per_request.

File: server/s3_get_object_action.cc [hunk 2]
```diff
@@ -356,11 +367,12 @@ void S3GetObjectAction::set_total_blocks_to_read_from_fragmented_object() {
...
           (extended_objects[first_byte_object_index].object_size -
            (first_byte_offset_to_read -
-            extended_objects[first_byte_object_index].start_offset_in_object));
+            extended_objects[first_byte_object_index].start_offset_in_object) +
+           1);
 
       extended_objects[last_byte_object_index].requested_object_size =
           last_byte_offset_to_read -
-          (extended_objects[last_byte_object_index].start_offset_in_object);
+          (extended_objects[last_byte_object_index].start_offset_in_object) + 1;
```
Comment (DPG17) [inline server/s3_get_object_action.cc:375]: While testing 12MB multpart object, in specific scenario (with 1 motr_units_per_request), get range was returning one bytes less than the expected. This code change fixes that problem.

**Comment by DPG17** [comment]
Pre-merge test passed http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/830/

**Comment by t7ko-seagate** [comment]
Deployment job succeeded:

http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/586/

**Review by t7ko-seagate — APPROVED** [review]

**Pull request merged**